site/app/libraries/homework/UseCases/LibraryAddUseCase.php

```diff
@@ -0,0 +1,126 @@
+<?php namespace app\libraries\homework\UseCases;
+
+use app\libraries\Core;
+use app\libraries\FileUtils;
+use app\libraries\homework\Gateways\LibraryGateway;
+use app\libraries\homework\Gateways\Library\LibraryGatewayFactory;
+
+class LibraryAddResponse {
+    /** @var string */
+    protected $message;
+
+    /** @var string */
+    public $error;
+
+    public function __construct(string $message = '') {
+        $this->message = $message;
+    }
+
+    public function getMessage(): string {
+        return $this->message;
+    }
+
+    public static function error(string $message) {
+        $response = new static;
+        $response->error = $message;
+        return $response;
+    }
+
+}
+
+class LibraryAddUseCase extends BaseUseCase {
+    /** @var LibraryGateway */
+    protected $gateway;
```

Comment:
Is this permitted by [Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html)? 

Does it break the Dependency Rule, "Nothing in an inner circle can know anything at all about something in an outer circle", to have a UseCase reference a Gateway?

Reply:
Yes, this is permitted becuase I am using [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle). Uncle bob talks about this in the link you mentioned under the section 'Crossing boundaries'.